File: ios/Porchivo/Views/LoginScreen.swift
//
//  LoginScreen.swift
//  Porchivo
//
//  Magic link authentication with the Porchivo welcome-porch hero illustration.
//  Email → 6-digit OTP code → verified. No passwords. A developer login link
//  sits under the magic link button for internal testing bypass.
//

import SwiftUI

struct LoginScreen: View {
    @Environment(AppState.self) private var appState
    @Environment(\.porchivo) private var c
    @Environment(\.colorScheme) private var colorScheme

    @State private var email = ""
    @State private var otpCode = ""
    @State private var phase: AuthPhase = .email
    @State private var isSubmitting = false
    @State private var linkSent = false
    @State private var showAuthFail = false
    @FocusState private var focusedField: FocusField?

    enum AuthPhase {
        case email
        case code
    }

    enum FocusField {
        case email
        case otp
    }

    var body: some View {
        if showAuthFail {
            AuthFailScreen(
                onBack: {
                    showAuthFail = false
                    phase = .email
                    otpCode = ""
                    appState.authError = nil
                },
                onCreateAccount: {
                    showAuthFail = false
                    phase = .email
                    otpCode = ""
                    appState.authError = nil
                }
            )
        } else {
        ZStack {
            // Hero illustration fills the background. In dark mode we dim it
            // slightly so the white porch remains readable without clashing.
            Image("LoginHero")
                .resizable()
                .aspectRatio(contentMode: .fill)
                .ignoresSafeArea()
                .overlay(colorScheme == .dark ? Color.black.opacity(0.28) : Color.clear)

            // Scrollable controls overlay the bottom half of the illustration.
            ScrollView {
                VStack(spacing: 0) {
                    Spacer().frame(height: 420)
                    controlsCard
                }
                .padding(.horizontal, 24)
                .padding(.bottom, 32)
            }
            .scrollDismissesKeyboard(.interactively)
        }
        .onAppear { focusedField = .email }
        }
    }

    // MARK: - Bottom controls card

    private var controlsCard: some View {
        VStack(spacing: 20) {
            switch phase {
            case .email:
                emailPhase
            case .code:
                codePhase
            }
            if let err = appState.authError {
                errorBanner(err)
            }
            if appState.isSupabaseConfigured == false {
                demoHint
            }
        }
        .padding(24)
        .background(
            .ultraThinMaterial,
            in: .rect(cornerRadius: Radius.xl)
        )
        .overlay(
            RoundedRectangle(cornerRadius: Radius.xl)
                .stroke(Color.white.opacity(0.35), lineWidth: 1)
        )
        .shadow(color: Color.black.opacity(0.12), radius: 24, x: 0, y: 12)
    }

    // MARK: - Email phase

    private var emailPhase: some View {
        VStack(spacing: 20) {
            emailField
            PrimaryButton(
                title: isSubmitting ? "Sending link…" : "Send magic link",
                systemImage: "envelope.fill",
                isLoading: isSubmitting,
                action: sendLink,
                enabled: isValidEmail && !isSubmitting
            )
            if linkSent {
                Text("Check your email for a 6-digit code.")
                    .font(.system(size: 12, weight: .medium))
                    .foregroundStyle(c.textSecondary)
                    .multilineTextAlignment(.center)
            }
            developerLoginLink
        }
    }

    private var emailField: some View {
        HStack(spacing: 10) {
            Image(systemName: "envelope.fill")
                .foregroundStyle(Color(hex: 0x8B5E3C))
            TextField("Enter your email", text: $email)
                .font(.system(size: 16))
                .foregroundStyle(Color(hex: 0x3D2B1F))
                .keyboardType(.emailAddress)
                .textInputAutocapitalization(.never)
                .autocorrectionDisabled()
                .focused($focusedField, equals: .email)
                .submitLabel(.go)
                .onSubmit { sendLink() }
        }
        .padding(.horizontal, 14)
        .padding(.vertical, 14)
        .background(Color(hex: 0xF5E6D3), in: .rect(cornerRadius: Radius.md))
        .overlay(
            RoundedRectangle(cornerRadius: Radius.md)
                .stroke(Color(hex: 0xD4A574), lineWidth: 1)
        )
    }

    // MARK: - Code phase

    private var codePhase: some View {
        VStack(spacing: 24) {
            Text("We sent a 6-digit code to \(email)")
                .font(.system(size: 13, weight: .medium))
                .foregroundStyle(c.textSecondary)
                .multilineTextAlignment(.center)
            otpInput
            PrimaryButton(
                title: isSubmitting ? "Verifying…" : "Verify code",
                systemImage: "checkmark.shield.fill",
                isLoading: isSubmitting,
                action: verifyCode,
                enabled: otpComplete && !isSubmitting
            )
            Button {
                Haptics.selection()
                withAnimation { phase = .email }
                otpCode = ""
                appState.authError = nil
            } label: {
                Text("Use a different email")
                    .font(.system(size: 14, weight: .semibold))
                    .foregroundStyle(c.accent)
            }
            Button {
                Haptics.selection()
                Task { await resendCode() }
            } label: {
                Text("Resend code")
                    .font(.system(size: 13, weight: .medium))
                    .foregroundStyle(c.textSecondary)
            }
        }
    }

    /// Single hidden TextField overlaid with visual digit cells.
    /// Uses `.textContentType(.oneTimeCode)` for iOS SMS/email autofill.
    private var otpInput: some View {
        ZStack {
            // Invisible TextField handles all keyboard input + autofill
            TextField("", text: $otpCode)
                .keyboardType(.numberPad)
                .textContentType(.oneTimeCode)
                .focused($focusedField, equals: .otp)
                .foregroundStyle(Color.clear)
                .background(.clear)
                .onChange(of: otpCode) { _, newValue in
                    // Filter to digits only, max 6
                    let filtered = newValue.filter { $0.isNumber }
                    otpCode = String(filtered.prefix(6))
                    // Auto-submit when 6 digits entered
                    if otpCode.count == 6 {
                        verifyCode()
                    }
                }
                .frame(width: 1, height: 1)
                .opacity(0.01)
                .allowsHitTesting(true)

            // Visual digit cells
            HStack(spacing: 10) {
                ForEach(0..<6, id: \.self) { idx in
                    let digit = idx < otpCode.count
                        ? String(otpCode[otpCode.index(otpCode.startIndex, offsetBy: idx)])
                        : ""
                    let isActive = idx == otpCode.count && focusedField == .otp
                    OTPDigitCell(digit: digit, isActive: isActive)
                }
            }
            .allowsHitTesting(false)
        }
        .contentShape(Rectangle())
        .onTapGesture {
            focusedField = .otp
        }
    }

    // MARK: - Developer bypass

    private var developerLoginLink: some View {
        Button {
            Haptics.selection()
            isSubmitting = true
            Task { @MainActor in
                appState.developerLogin()
                isSubmitting = false
            }
        } label: {
            Text("Developer login")
                .font(.system(size: 13, weight: .medium))
                .foregroundStyle(c.textSecondary)
                .underline()
        }
        .disabled(isSubmitting)
    }

    // MARK: - Error + demo

    private func errorBanner(_ msg: String) -> some View {
        Text(msg)
            .font(.system(size: 13, weight: .medium))
            .foregroundStyle(c.danger)
            .multilineTextAlignment(.center)
            .frame(maxWidth: .infinity)
            .padding(.horizontal, 12)
            .padding(.vertical, 10)
            .background(c.dangerSoft, in: .rect(cornerRadius: Radius.sm))
    }

    private var demoHint: some View {
        Text("Demo mode — backend not configured.\nEnter any email, then any 6 digits to sign in.")
            .font(.system(size: 11, weight: .medium))
            .foregroundStyle(c.textMuted)
            .multilineTextAlignment(.center)
    }

    // MARK: - Actions

    private var isValidEmail: Bool {
        let trimmed = email.trimmingCharacters(in: .whitespaces)
        return trimmed.contains("@") && trimmed.contains(".")
    }

    private var otpComplete: Bool {
        otpCode.count == 6
    }

    private func sendLink() {
        guard isValidEmail else { return }
        Haptics.light()
        isSubmitting = true
        let emailCopy = email
        Task { @MainActor in
            defer { isSubmitting = false }
            let ok = await appState.sendMagicLink(email: emailCopy)
            if ok {
                withAnimation {
                    phase = .code
                    linkSent = true
                    appState.authError = nil
                }
                focusedField = .otp
            } else {
                // authError is already set by AppState with the real Supabase error.
                // Only set a fallback if AppState didn't populate one.
                if appState.authError == nil {
                    appState.authError = "Could not send magic link. Check your email and try again."
                }
            }
        }
    }

    private func verifyCode() {
        guard otpComplete else { return }
        Haptics.light()
        isSubmitting = true
        let emailCopy = email
        let codeCopy = otpCode
        Task { @MainActor in
            defer { isSubmitting = false }
            let ok = await appState.verifyOtp(email: emailCopy, token: codeCopy)
            if ok {
                Haptics.success()
            } else {
                // If the error suggests the user doesn't have an account,
                // show the oops screen instead of an inline error.
                let err = appState.authError ?? ""
                if err.lowercased().contains("invalid") || err.lowercased().contains("expired") || err.lowercased().contains("code") {
                    // Wrong/expired code — keep inline error
                } else {
                    showAuthFail = true
                }
            }
            // On success, RootView picks up authState change automatically.
            // On failure, authError is set by AppState.
        }
    }

    private func resendCode() async {
        isSubmitting = true
        defer { isSubmitting = false }
        let ok = await appState.sendMagicLink(email: email)
        if !ok && appState.authError == nil {
            appState.authError = "Could not resend code. Try again."
        }
    }
}

// MARK: - OTP digit cell (display only)

private struct OTPDigitCell: View {
    @Environment(\.porchivo) private var c
    let digit: String
    let isActive: Bool

    var body: some View {
        Text(digit)
            .font(.system(size: 24, weight: .bold, design: .rounded))
            .foregroundStyle(c.textPrimary)
            .frame(width: 48, height: 56)
            .background(c.surface, in: .rect(cornerRadius: Radius.md))
            .overlay {
                RoundedRectangle(cornerRadius: Radius.md)
                    .stroke(isActive ? c.accent : c.border, lineWidth: isActive ? 2 : 1)
                if isActive {
                    // Caret indicator on the active cell
                    Rectangle()
                        .fill(c.accent)
                        .frame(width: 2, height: 24)
                }
            }
            .animation(.easeInOut(duration: 0.2), value: isActive)
    }
}

#Preview {
    LoginScreen()
        .environment(AppState())
}

File: android/app/src/main/java/com/rork/porchivo/ui/screens/LoginScreen.kt
package com.rork.porchivo.ui.screens

import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Mail
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.OutlinedTextFieldDefaults
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import kotlinx.coroutines.launch
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextDecoration
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import androidx.lifecycle.viewmodel.compose.viewModel
import com.rork.porchivo.R
import com.rork.porchivo.data.AuthState
import com.rork.porchivo.ui.theme.PorchivoTheme
import com.rork.porchivo.ui.viewmodel.AppViewModel

/**
 * Auth screen — magic link authentication with the Porchivo welcome-porch hero.
 * Email → 6-digit OTP code → verified. No passwords.
 * A developer login link sits under the magic link button for internal testing bypass.
 */
@Composable
fun LoginScreen(
    onAuthSuccess: () -> Unit,
    modifier: Modifier = Modifier,
    appViewModel: AppViewModel = viewModel(),
) {
    val c = PorchivoTheme.colors
    val authState by appViewModel.authState.collectAsStateWithLifecycle()
    val authError by appViewModel.authError.collectAsStateWithLifecycle()

    var email by remember { mutableStateOf("") }
    var otpCode by remember { mutableStateOf("") }
    var phase by remember { mutableStateOf(AuthPhase.EMAIL) }
    var linkSent by remember { mutableStateOf(false) }

    val isLoading = authState is AuthState.Loading
    val scope = rememberCoroutineScope()

    // Side-effect: navigate out once authenticated.
    LaunchedEffect(authState) {
        if (authState is AuthState.Authenticated) {
            onAuthSuccess()
        }
    }

    Box(modifier = modifier.fillMaxSize()) {
        // Hero illustration fills the background.
        Image(
            painter = painterResource(id = R.drawable.porchlogin),
            contentDescription = null,
            modifier = Modifier.fillMaxSize(),
            contentScale = ContentScale.Crop,
        )

        // Dark scrim so the white porch doesn't blow out in dark mode.
        Box(
            modifier = Modifier
                .fillMaxSize()
                .alpha(0.22f)
                .background(Color.Black),
        )

        // Scrollable bottom controls.
        Column(
            modifier = Modifier
                .fillMaxSize()
                .verticalScroll(rememberScrollState())
                .padding(horizontal = 24.dp, vertical = 32.dp),
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.Bottom,
        ) {
            Spacer(modifier = Modifier.weight(1f))
            ControlsCard(
                email = email,
                onEmailChange = { email = it },
                otpCode = otpCode,
                onOtpChange = { otpCode = it.filter(Char::isDigit).take(6) },
                phase = phase,
                onPhaseChange = { phase = it },
                linkSent = linkSent,
                onLinkSentChange = { linkSent = it },
                authError = authError,
                isLoading = isLoading,
                onSendMagicLink = {
                    scope.launch {
                        val ok = appViewModel.sendMagicLink(email)
                        if (ok) {
                            linkSent = true
                            phase = AuthPhase.CODE
                        }
                    }
                },
                onVerifyOtp = {
                    scope.launch { appViewModel.verifyOtp(email, otpCode) }
                },
                onDeveloperLogin = { appViewModel.developerLogin() },
            )
        }
    }
}

@Composable
private fun ControlsCard(
    email: String,
    onEmailChange: (String) -> Unit,
    otpCode: String,
    onOtpChange: (String) -> Unit,
    phase: AuthPhase,
    onPhaseChange: (AuthPhase) -> Unit,
    linkSent: Boolean,
    onLinkSentChange: (Boolean) -> Unit,
    authError: String?,
    isLoading: Boolean,
    onSendMagicLink: () -> Unit,
    onVerifyOtp: () -> Unit,
    onDeveloperLogin: () -> Unit,
) {
    val c = PorchivoTheme.colors

    Column(
        modifier = Modifier
            .fillMaxWidth()
            .background(c.surface.copy(alpha = 0.92f), RoundedCornerShape(24.dp))
            .padding(24.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
    ) {
        when (phase) {
            AuthPhase.EMAIL -> EmailPhase(
                email = email,
                onEmailChange = onEmailChange,
                linkSent = linkSent,
                authError = authError,
                isLoading = isLoading,
                onSendMagicLink = onSendMagicLink,
                onDeveloperLogin = onDeveloperLogin,
            )
            AuthPhase.CODE -> CodePhase(
                email = email,
                otpCode = otpCode,
                onOtpChange = onOtpChange,
                authError = authError,
                isLoading = isLoading,
                onVerifyOtp = onVerifyOtp,
                onBackToEmail = { onPhaseChange(AuthPhase.EMAIL) },
                onResendCode = onSendMagicLink,
            )
        }
    }
}

@Composable
private fun EmailPhase(
    email: String,
    onEmailChange: (String) -> Unit,
    linkSent: Boolean,
    authError: String?,
    isLoading: Boolean,
    onSendMagicLink: () -> Unit,
    onDeveloperLogin: () -> Unit,
) {
    val c = PorchivoTheme.colors
    val isValidEmail = email.trim().contains("@") && email.trim().contains(".")

    val fieldColors = OutlinedTextFieldDefaults.colors(
        focusedBorderColor = Color(0xFFD4A574),
        unfocusedBorderColor = Color(0xFFD4A574),
        focusedContainerColor = Color(0xFFF5E6D3),
        unfocusedContainerColor = Color(0xFFF5E6D3),
        focusedTextColor = Color(0xFF3D2B1F),
        unfocusedTextColor = Color(0xFF3D2B1F),
        cursorColor = Color(0xFF8B5E3C),
    )

    OutlinedTextField(
        value = email,
        onValueChange = onEmailChange,
        placeholder = { Text("Enter your email", color = Color(0xFF8B5E3C)) },
        leadingIcon = {
            Icon(
                imageVector = Icons.Filled.Mail,
                contentDescription = null,
                tint = Color(0xFF8B5E3C),
                modifier = Modifier.width(20.dp),
            )
        },
        colors = fieldColors,
        shape = RoundedCornerShape(12.dp),
        singleLine = true,
        keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Email),
        modifier = Modifier.fillMaxWidth(),
    )

    Spacer(modifier = Modifier.height(16.dp))

    Button(
        onClick = onSendMagicLink,
        enabled = isValidEmail && !isLoading,
        modifier = Modifier
            .fillMaxWidth()
            .height(52.dp),
        shape = RoundedCornerShape(16.dp),
        colors = ButtonDefaults.buttonColors(
            containerColor = Color(0xFF2E5A96),
            disabledContainerColor = c.elevated,
        ),
    ) {
        if (isLoading) {
            CircularProgressIndicator(
                color = c.onAccent,
                strokeWidth = 2.dp,
                modifier = Modifier.width(20.dp),
            )
        } else {
            Icon(
                imageVector = Icons.Filled.Mail,
                contentDescription = null,
                tint = c.onAccent,
                modifier = Modifier.width(20.dp),
            )
            Spacer(modifier = Modifier.width(8.dp))
            Text(
                text = if (linkSent) "Sending link…" else "Send magic link",
                fontSize = 16.sp,
                fontWeight = FontWeight.Bold,
                color = c.onAccent,
            )
        }
    }

    if (linkSent) {
        Spacer(modifier = Modifier.height(8.dp))
        Text(
            text = "Check your email for a 6-digit code.",
            color = c.textSecondary,
            fontSize = 12.sp,
            fontWeight = FontWeight.Medium,
            textAlign = TextAlign.Center,
        )
    }

    Spacer(modifier = Modifier.height(12.dp))

    TextButton(onClick = onDeveloperLogin) {
        Text(
            text = "Developer login",
            color = c.textSecondary,
            fontSize = 13.sp,
            fontWeight = FontWeight.Medium,
            textDecoration = TextDecoration.Underline,
        )
    }

    authError?.let { errorMsg ->
        Spacer(modifier = Modifier.height(12.dp))
        Text(
            text = errorMsg,
            color = c.danger,
            fontSize = 13.sp,
            fontWeight = FontWeight.Medium,
            textAlign = TextAlign.Center,
            modifier = Modifier.fillMaxWidth(),
        )
    }
}

@Composable
private fun CodePhase(
    email: String,
    otpCode: String,
    onOtpChange: (String) -> Unit,
    authError: String?,
    isLoading: Boolean,
    onVerifyOtp: () -> Unit,
    onBackToEmail: () -> Unit,
    onResendCode: () -> Unit,
) {
    val c = PorchivoTheme.colors
    val isOtpComplete = otpCode.length == 6

    Text(
        text = "We sent a 6-digit code to $email",
        color = c.textSecondary,
        fontSize = 13.sp,
        fontWeight = FontWeight.Medium,
        textAlign = TextAlign.Center,
        modifier = Modifier.fillMaxWidth(),
    )

    Spacer(modifier = Modifier.height(16.dp))

    OutlinedTextField(
        value = otpCode,
        onValueChange = onOtpChange,
        placeholder = { Text("6-digit code", color = c.textMuted) },
        colors = OutlinedTextFieldDefaults.colors(
            focusedBorderColor = c.accent,
            unfocusedBorderColor = c.border,
            focusedContainerColor = c.surface,
            unfocusedContainerColor = c.surface,
            focusedTextColor = c.textPrimary,
            unfocusedTextColor = c.textPrimary,
            cursorColor = c.accent,
        ),
        shape = RoundedCornerShape(12.dp),
        singleLine = true,
        keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.NumberPassword),
        modifier = Modifier.fillMaxWidth(),
    )

    Spacer(modifier = Modifier.height(16.dp))

    Button(
        onClick = onVerifyOtp,
        enabled = isOtpComplete && !isLoading,
        modifier = Modifier
            .fillMaxWidth()
            .height(52.dp),
        shape = RoundedCornerShape(16.dp),
        colors = ButtonDefaults.buttonColors(
            containerColor = Color(0xFF2E5A96),
            disabledContainerColor = c.elevated,
        ),
    ) {
        if (isLoading) {
            CircularProgressIndicator(
                color = c.onAccent,
                strokeWidth = 2.dp,
                modifier = Modifier.width(20.dp),
            )
        } else {
            Text(
                text = "Verify code",
                fontSize = 16.sp,
                fontWeight = FontWeight.Bold,
                color = c.onAccent,
            )
        }
    }

    Spacer(modifier = Modifier.height(8.dp))

    TextButton(onClick = onBackToEmail) {
        Text(
            text = "Use a different email",
            color = c.accent,
            fontSize = 14.sp,
            fontWeight = FontWeight.SemiBold,
        )
    }

    TextButton(onClick = onResendCode) {
        Text(
            text = "Resend code",
            color = c.textSecondary,
            fontSize = 13.sp,
            fontWeight = FontWeight.Medium,
        )
    }

    authError?.let { errorMsg ->
        Spacer(modifier = Modifier.height(8.dp))
        Text(
            text = errorMsg,
            color = c.danger,
            fontSize = 13.sp,
            fontWeight = FontWeight.Medium,
            textAlign = TextAlign.Center,
            modifier = Modifier.fillMaxWidth(),
        )
    }
}

private enum class AuthPhase {
    EMAIL,
    CODE,
}
